### DISCUSSION TITLE
Set theme of windows app title bar

### DISCUSSION THREAD

**Discussion by skanvk15** [microsoft/WindowsAppSDK#2677] [General]
I have C# WinUI 3 app (output type: Windows Application, Target framework: .NET 6.0)  and am setting the theme programmatically using the RequestedTheme property

```
        private void LigthButton_Click(object sender, RoutedEventArgs e)
        {
            MainGrid.RequestedTheme = ElementTheme.Light;
        }

        private void DarkButton_Click(object sender, RoutedEventArgs e)
        {
            MainGrid.RequestedTheme = ElementTheme.Dark;
        }
```

This works for all areas except the title bar. How do I change the window TitleBar theme? 

![image](https://user-images.githubusercontent.com/51756883/175667321-feddf915-4b37-42d4-9a53-1ecd3de639ca.png)

**Comment by ghost1372** [comment]
Titlebar theme will not change, i reported this issue before and until today not fixed.
But you can fix this if you extend your appwindow to titlebar
```appWindow.Titlebar.ExtendsContentIntoTitleBar =  true;```

**Reply by skanvk15** [reply]
I tried this suggestion but the buttons on the right (minimize, maximize/restore, close) don't change. I want to keep those buttons and have the title bar use the selected theme.

But the suggestion is a good answer if you don't need the default title bar.

**Reply by ghost1372** [reply]
you need to update buttons background.
you can use my [nuget package](https://github.com/ghost1372/SettingsUI) and **ThemeHelper** class for handle this issues.
or you can copy and paste ThemeHelper class
https://github.com/ghost1372/SettingsUI/blob/main/src/SettingsUI/Tools/Helpers/ThemeHelper.cs

**Reply by sigmarsson** [reply]
Is the nuget package still existing ?

**Reply by ghost1372** [reply]
> Is the nuget package still existing ?

Yes, moved to a new package and renamed to ThemeService.
https://github.com/WinUICommunity/WinUICommunity

**Reply by sigmarsson** [reply]
@ghost1372 - Thank you for your answer!

**Reply by sigmarsson** [reply]
@ghost1372 Where is the method WindowHelper.GetRasterizationScaleForElement ?

**Reply by ghost1372** [reply]
> @ghost1372 Where is the method WindowHelper.GetRasterizationScaleForElement ?

See docs
https://ghost1372.github.io/winUICommunityCore/helpers/generalHelper/index.html?highlight=rast#GetRasterizationScaleForElement

**Reply by sigmarsson** [reply]
@ghost1372 - Thank you. After calling the ThemeService, is there anything to call the title bar reflects to the new theme ? 

`_themeService.SetElementTheme(ElementTheme.Default);`

For example [this method](https://github.com/microsoft/WinUI-Gallery/blob/5079f03292c29ceaaa2aab1bf89619599f4989d6/WinUIGallery/ControlPages/TitleBarPage.xaml.cs#L148)

**Reply by ghost1372** [reply]
@sigmarsson  you should first use : `Window.ExtendContentIntoTitleBar = true` or `Window.AppWindow.TitleBar.ExtendsContentIntoTitleBar = true` both are same effect, but if you use second option, you can use ThemeService [ConfigTitleBar](https://ghost1372.github.io/winUICommunityCore/themeService/#ConfigTitleBar)

**Reply by sigmarsson** [reply]
@ghost1372 - Thank you. Is there a way to paint the title bar itself too ? for me the 3 buttons are being repainted only. 

![image](https://github.com/microsoft/WindowsAppSDK/assets/43340959/50e156f1-5b9c-4e87-bbf6-aae97ae70965)

**Reply by ghost1372** [reply]
you can, if you use `AppWindow.TitleBar.Extend,....` you can use [ConfigTitleBar](https://ghost1372.github.io/winUICommunityCore/themeService/#ConfigTitleBar)

**Reply by sigmarsson** [reply]
@ghost1372 let's say, yes it is working until after I set a `<Grid x:Name="TitleBar"/>` with `SetTitleBar` in the shell view's `Loaded` handler. In thhis case the title bar remains always the OS default.

**Reply by ghost1372** [reply]
i did not test this situation, but i suggest you to use our [TitleBar](https://ghost1372.github.io/winUICommunityComponents/controls/titleBar/) control:
which is working fine

**Comment by castorix** [comment]
On Windows 11, there is also DWMWA_CAPTION_COLOR/DWMWA_TEXT_COLOR to set colors (cannot test, I have only Windows 10)

**Comment by sigmarsson** [comment]
@castorix could you provide the complete code please ? I invoke this beenath on 11 but to no avail. 


```
using System.Runtime.InteropServices;

using Microsoft.UI;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Media;

using Windows.UI;

using System;

namespace Weather.History.Windows;

// Helper class to workaround custom title bar bugs.
// DISCLAIMER: The resource key names and color values used below are subject to change. Do not depend on them.
// https://github.com/microsoft/TemplateStudio/issues/4516
internal class Win32Ex
{
    private const int WAINACTIVE = 0x00;
    private const int WAACTIVE = 0x01;
    private const int WMACTIVATE = 0x0006;

    [DllImport("user32.dll")]
    private static extern IntPtr GetActiveWindow();

    [DllImport("user32.dll", CharSet = CharSet.Auto)]
    private static extern IntPtr SendMessage(IntPtr hWnd, int msg, int wParam, IntPtr lParam);

    public static void UpdateTitleBar(UIElement shell)
    {
        if (shell is FrameworkElement frameworkElement)
        {
            UpdateTitleBar(frameworkElement.RequestedTheme);
        }
        else
        {
            throw new ArgumentException($"Unable to update application title bar: {shell.GetType()}");
        }
    }

    public static void UpdateTitleBar(ElementTheme theme)
    {
        if (App.Window.ExtendsContentIntoTitleBar)
        {
            if (theme != ElementTheme.Default)
            {
                Application.Current.Resources["WindowCaptionForeground"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Colors.White),
                    ElementTheme.Light => new SolidColorBrush(Colors.Black),
                    _ => new SolidColorBrush(Colors.Transparent)
                };

                Application.Current.Resources["WindowCaptionForegroundDisabled"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Color.FromArgb(0x66, 0xFF, 0xFF, 0xFF)),
                    ElementTheme.Light => new SolidColorBrush(Color.FromArgb(0x66, 0x00, 0x00, 0x00)),
                    _ => new SolidColorBrush(Colors.Transparent)
                };

                Application.Current.Resources["WindowCaptionButtonBackgroundPointerOver"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Color.FromArgb(0x33, 0xFF, 0xFF, 0xFF)),
                    ElementTheme.Light => new SolidColorBrush(Color.FromArgb(0x33, 0x00, 0x00, 0x00)),
                    _ => new SolidColorBrush(Colors.Transparent)
                };

                Application.Current.Resources["WindowCaptionButtonBackgroundPressed"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Color.FromArgb(0x66, 0xFF, 0xFF, 0xFF)),
                    ElementTheme.Light => new SolidColorBrush(Color.FromArgb(0x66, 0x00, 0x00, 0x00)),
                    _ => new SolidColorBrush(Colors.Transparent)
                };

                Application.Current.Resources["WindowCaptionButtonStrokePointerOver"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Colors.White),
                    ElementTheme.Light => new SolidColorBrush(Colors.Black),
                    _ => new SolidColorBrush(Colors.Transparent)
                };

                Application.Current.Resources["WindowCaptionButtonStrokePressed"] = theme switch
                {
                    ElementTheme.Dark => new SolidColorBrush(Colors.White),
                    ElementTheme.Light => new SolidColorBrush(Colors.Black),
                    _ => new SolidColorBrush(Colors.Transparent)
                };
            }

            Application.Current.Resources["WindowCaptionBackground"] = new SolidColorBrush(Colors.Transparent);
            Application.Current.Resources["WindowCaptionBackgroundDisabled"] = new SolidColorBrush(Colors.Transparent);

            var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(App.Window);

            if (hwnd == GetActiveWindow())
            {
                SendMessage(hwnd, WMACTIVATE, WAINACTIVE, IntPtr.Zero);
                SendMessage(hwnd, WMACTIVATE, WAACTIVE, IntPtr.Zero);
            }
            else
            {
                SendMessage(hwnd, WMACTIVATE, WAACTIVE, IntPtr.Zero);
                SendMessage(hwnd, WMACTIVATE, WAINACTIVE, IntPtr.Zero);
            }
        }
    }
}

```

**Reply by DarranRowe** [reply]
If you are on 1.4, try using [AppWindowTitleBar](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindowtitlebar?view=windows-app-sdk-1.4) first.
[Window](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.window?view=windows-app-sdk-1.4) has the AppWindow property which allows you to get the underlying [AppWindow](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.windowing.appwindow?view=windows-app-sdk-1.4) instance. AppWindow has the TitleBar property which gives you the AppWindowTitleBar instance.
I mention this because the [1.4 release notes](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel#known-issues) states that the custom title bar is now drawn by AppWindow. This would mean that AppWindowTitleBar would be used to update the button colours for the AppWindow's version of the custom title bar.

AppWindowTitleBar should also respond properly to DwmSetWindowAttribute with DWMWA_USE_IMMERSIVE_DARK_MODE. I have noticed that you sometimes have to update a property to refresh the colours. So I normally just get one of the properties and use that to set the same property again.

**Reply by tpoint75** [reply]
Thank you very much! It is so hard to find this for winrt/c++. The documentation is c# always..

AppWindow().TitleBar().ButtonForegroundColor(Colors::White());

**Comment by sigmarsson** [comment]
@DarranRowe Thank you. How shall I set the light titlebar from dark with `DwmSetWindowAttribute` ?

![image](https://github.com/microsoft/WindowsAppSDK/assets/43340959/6b1a6507-adc2-4163-9d2c-9e671e315054)

**Comment by sigmarsson** [comment]
If I follow your instructions the effect of  `ExtendsContentIntoTitleBar` lost ;

![image](https://github.com/microsoft/WindowsAppSDK/assets/43340959/19c1c3c1-4aa8-4a9c-8ac9-9ab04d8c760b)


`Window.AppWindow.TitleBar.BackgroundColor = Colors.Green;`

**Reply by DarranRowe** [reply]
From my previous post:

>This would mean that AppWindowTitleBar would be used to update the **button colours** for the AppWindow's version of the custom title bar.

I would claim that I only instructed you to modify the properties related to the buttons. But that could be easy to miss, especially if you are looking for a direct replacement to the Xaml provided method in 1.3 or earlier.

Currently, the entire functionality around Microsoft.UI.Xaml.Window.SetTitleBar does not work. This means that for the caption area itself, you would have to set the Foreground and Background of the individual controls themselves.

Although I would expect AppWindowTitleBar to provide some form of error when you attempt to use the functionality only usable in simple customisation instead of just disabling the custom title bar.

**Reply by sigmarsson** [reply]
Will the `SetTitleBar` work again with 1.5 ?
